src/storage-client/src/controller.rs

```diff
@@ -2218,31 +2239,46 @@ where
             }
             Some(StorageResponse::StatisticsUpdates(source_stats, sink_stats)) => {
                 // Note we only hold the locks while moving some plain-old-data around here.
-
-                let mut shared_stats = self.state.source_statistics.lock().expect("poisoned");
-                for stat in source_stats {
-                    let shared_stats = shared_stats.entry(stat.id).or_default();
-                    // We just write the whole object, as the update from storage represents the
-                    // current values.
-                    shared_stats.insert(stat.worker_id, stat);
+                //
+                // We just write the whole object, as the update from storage represents the
+                // current values.
+                //
+                // We don't overwrite removed objects, as we may have received a late
+                // `StatisticsUpdates` while we were shutting down the storage object.
+                {
+                    let mut shared_stats = self.state.source_statistics.lock().expect("poisoned");
+                    for stat in source_stats {
+                        statistics::StatsInitState::set_if_not_removed(
+                            shared_stats.get_mut(&stat.id),
+                            stat.worker_id,
+                            stat,
+                        )
+                    }
                 }
 
-                let mut shared_stats = self.state.sink_statistics.lock().expect("poisoned");
-                for stat in sink_stats {
-                    let shared_stats = shared_stats.entry(stat.id).or_default();
-                    // We just write the whole object, as the update from storage represents the
-                    // current values.
-                    shared_stats.insert(stat.worker_id, stat);
+                {
+                    let mut shared_stats = self.state.sink_statistics.lock().expect("poisoned");
+                    for stat in sink_stats {
+                        statistics::StatsInitState::set_if_not_removed(
+                            shared_stats.get_mut(&stat.id),
+                            stat.worker_id,
+                            stat,
+                        )
+                    }
                 }
             }
         }
 
         // IDs of sources that were dropped whose statuses should be updated.
         let mut pending_source_drops = vec![];
 
-        // IDs of sinks that were dropped whose statuses should be updated.
+        // IDs of sinks that were dropped whose statuses should be updated (and statistics
+        // cleared).
         let mut pending_sink_drops = vec![];
```

Comment:
The unparallel structure of handling source and sink drops could use a comment; it's not totally clear to me why sink statistics exist only when there's a cluster while sources are different.